sys/net/gnrc/application_layer/tftp/gnrc_tftp.c

```diff
@@ -0,0 +1,848 @@
+    uint16_t mtu = _tftp_get_maximum_block_size();
+    if (_tftp_set_opts(&ctxt, mtu, 1, total_size) != FINISHED) {
+        return -EINVAL;
+    }
+
+    /* start the process */
+    return _tftp_do_client_transfer(&ctxt);
+}
+
+#define INT_DIGITS 6
+char *itoa(int i, size_t *len) {
+    /* Room for INT_DIGITS digits, - and '\0' */
+    static char buf[INT_DIGITS + 2];
+    char *p = buf + INT_DIGITS + 1; /* points to terminating '\0' */
+    if (i >= 0) {
+        do {
+            *--p = '0' + (i % 10);
+            i /= 10;
+        } while (i != 0);
+        *len = (INT_DIGITS + 1) - (p - buf);
+        return p;
+    }
+
+    return NULL;
+}
+#undef INT_DIGITS
+
+int _tftp_init_ctxt(ipv6_addr_t *addr, const char *file_name,
+                    tftp_transfer_start_callback start_cb,
+                    tftp_data_callback data_cb, tftp_opcodes_t op,
+                    tftp_context_t *ctxt) {
+
+    if (!addr || !data_cb) {
+        return FAILED;
+    }
+
+    memset(ctxt, 0, sizeof(tftp_context_t));
+
+    /* set the default context parameters */
+    ctxt->op = op;
+    ctxt->peer = addr;
+    ctxt->mode = OCTET;
+    ctxt->data_cb = data_cb;
+    ctxt->start_cb = start_cb;
+    strncpy(ctxt->file_name, file_name, GNRC_TFTP_MAX_FILENAME_LEN);
+    ctxt->file_name[GNRC_TFTP_MAX_FILENAME_LEN - 1] = 0;
+    ctxt->dst_port = GNRC_TFTP_DEFAULT_DST_PORT;
+
+    /* transport layer parameters */
+    ctxt->block_size = GNRC_TFTP_MAX_TRANSFER_UNIT;
+    ctxt->block_timeout = 1;
+
+    /* generate a random source UDP source port */
+    do {
+        ctxt->src_port = (genrand_uint32() & 0xff) + GNRC_TFTP_DEFAULT_SRC_PORT;
+    } while (gnrc_netreg_num(GNRC_NETTYPE_UDP, ctxt->src_port));
+
+    return FINISHED;
+}
+
+void _tftp_set_default_options(tftp_context_t *ctxt) {
+    ctxt->block_size = GNRC_TFTP_MAX_TRANSFER_UNIT;
+    ctxt->timeout = 1;
+    ctxt->transfer_size = 0;
+    ctxt->use_options = false;
+}
+
+int _tftp_set_opts(tftp_context_t *ctxt, size_t blksize, uint16_t timeout, size_t total_size) {
+    if (blksize > GNRC_TFTP_MAX_TRANSFER_UNIT || !timeout) {
+        return FAILED;
+    }
+
+    ctxt->block_size = blksize;
+    ctxt->timeout = timeout;
+    ctxt->block_timeout = timeout;
+    ctxt->transfer_size = total_size;
+    ctxt->use_options = true;
+
+    return FINISHED;
+}
+
+int _tftp_server(tftp_context_t *ctxt) {
+    msg_t msg;
+    gnrc_netreg_entry_t entry = { NULL, GNRC_TFTP_DEFAULT_DST_PORT,
+                              thread_getpid() };
+
+    while (1) {
+        int ret = BUSY;
+        bool got_client = false;
+
+        /* register our TFTP response listener */
+        if (gnrc_netreg_register(GNRC_NETTYPE_UDP, &entry)) {
+            DEBUG("tftp: error starting server.");
+            return FAILED;
+        }
+
+        /* main processing loop */
+        while (ret == BUSY) {
+            /* wait for a message */
+            msg_receive(&msg);
+            ret = _tftp_state_processes(ctxt, &msg);
```

Comment:
The `gnrc_pktsnip_t *pkt = m->content.ptr;` possibly contained in `msg` should be released here. I guess it's easier, code-wise, to release here instead of infront of every return in `_tftp_state_processes()`.